src/App.vue

```diff
@@ -0,0 +1,577 @@
+}
+
+export type Shader = {
+    succ: true,
+    code: string,
+    layout: Bindings,
+    hashedStrings: any,
+    reflection: ReflectionJSON,
+    threadGroupSize: ThreadGroupSize | { x: number, y: number, z: number }
+};
+
+export type MaybeShader = Shader | {
+    succ: false
+};
+
+function compileShader(userSource: string, entryPoint: string, compileTarget: typeof compileTargets[number], includePlaygroundModule = true): MaybeShader {
+    if (compiler == null) throw new Error("No compiler available");
+    const compiledResult = compiler.compile(userSource, entryPoint, compileTarget);
+    diagnosticsArea.value?.setEditorValue(compiler.diagnosticsMsg, true);
+
+    // If compile is failed, we just clear the codeGenArea
+    if (!compiledResult) {
+        codeGenArea.value?.setEditorValue('Compilation returned empty result.');
+        return { succ: false };
+    }
+
+    let [compiledCode, layout, hashedStrings, reflectionJsonObj, threadGroupSize] = compiledResult;
+    reflectionJson = reflectionJsonObj;
+
+    codeGenArea.value?.setEditorValue(compiledCode);
+    codeGenArea.value?.setLanguage(targetLanguageMap[compileTarget]);
+
+    // Update reflection info.
+    window.$jsontree.setJson("reflectionDiv", reflectionJson);
+    window.$jsontree.refreshAll();
+
+    return { succ: true, code: compiledCode, layout: layout, hashedStrings: hashedStrings, reflection: reflectionJson, threadGroupSize: threadGroupSize };
+}
+
+function restoreSelectedTargetFromURL() {
+    const urlParams = new URLSearchParams(window.location.search);
+    const target = urlParams.get('target');
+    if (target) {
+        if (!compileTargets.includes(target as any)) {
+            diagnosticsArea.value?.setEditorValue("Invalid target specified in URL: " + target);
+            return;
+        }
+        targetSelect.value!.setValue(target as any);
+        updateProfileOptions();
+    }
+}
+
+function restoreDemoSelectionFromURL() {
+    const urlParams = new URLSearchParams(window.location.search);
+    let demo = urlParams.get('demo');
+    if (demo) {
+        if (!demo.endsWith(".slang"))
+            demo += ".slang";
+        selectedDemo.value = demo;
+        loadDemo(demo);
+        return true;
+    }
+    return false;
+}
+
+async function runIfFullyInitialized() {
+    if (compiler && slangd && pageLoaded && codeEditor.value) {
+        (await import("./language-server")).initLanguageServer();
+
+        initialized.value = true;
+
+        restoreSelectedTargetFromURL();
+
+        if (restoreDemoSelectionFromURL()) { }
+        else if (codeEditor.value.getValue() == "") {
+            loadDemo(defaultShaderURL);
+        }
+        else {
+            compileOrRun();
+        }
+    }
+}
+
+function logError(message: string) {
+    diagnosticsArea.value?.appendEditorValue(message + "\n", true);
+}
+</script>
+
+<template>
+    <Tooltip ref="tooltip"></Tooltip>
+    <Transition>
+        <div id="loading-screen" v-if="!initialized">
```

Comment:
This is also defined in index.html, do we need both?

Reply:
The definition in index.html gets overridden. It's used to give users an earlier first render before the Vue code loads.

Reply:
I see. Maybe we can separate it out to slits own file that gets included from here and the vue template?